### PR TITLE
Nav until text

### PR DESCRIPTION
**Pull request by agrojean-ledger**

### CODE REVIEW

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [inline src/ragger/navigator/navigator.py:323]: Not sure this can work in use cases where no APDU is sent and no screen transition is triggered automatically.

It's probably going to result in a timeout error

File: src/ragger/backend/speculos.py
Comment (xchapron-ledger) [inline src/ragger/backend/speculos.py:155]: Add a `time.sleep(x)` here to give some time to other threads?

File: src/ragger/backend/speculos.py
Comment (xchapron-ledger) [inline src/ragger/backend/speculos.py:161]: `None`?

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [reply src/ragger/navigator/navigator.py:323]: On cureent SpeculosBackend implem it should immediatly return without timeout error as you start without context.

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:329]: However this one will probably raised in case `navigate_until_text` is called after sending an APDU which flow is quickly displayed.
Something like :
- Send APDU
- Flow displayed
- `navigate_until_text`:
    - first wait_for_screen_change(2.0) => get first screen of flow content
    - wait_for_screen_change(2.0) in the while loop => timeout as no navigation instruction was sent.

File: src/ragger/backend/speculos.py
Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:155]: I don't understand what would be the purpose of the sleep

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py]: I moved the navigation instruction above the wait for screen change in the loop. It should solve the problem you describe (and other use cases with no initial automatic screen change).

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py]: Added None as explicit return type.

File: src/ragger/backend/speculos.py
Comment (xchapron-ledger) [reply src/ragger/backend/speculos.py:155]: sorry, I put in on the wrong line.
I mean putting a sleep in the while loop to avoid polling `get_screen_content()` to fast.
I guess 0.1s between each poll won't slow tests?

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:331]: Timeout could be done after the comparison?
And maybe you can use `time()` directly instead of in a temporary variable?

**Comment by xchapron-ledger** [comment]
This misses some prototypes in backend/interface.py and a short documentation in readme.md.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:336]: Done.

File: src/ragger/backend/speculos.py
Comment (xchapron-ledger) [inline src/ragger/backend/speculos.py:170]:
```suggestion
            content = self._client.get_screen_content()
            while content == context:
                sleep(0.1)
                now = time()
                if (now - start > timeout):
                    raise TimeoutError("Timeout waiting for screen change")
                content = self._client.get_screen_content()
            return content
```
Pour éviter d'attendre si ça ne sert à rien ?

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:324]: not sure about this one. Shouldn't it be transparent as in `navigate_and_compare`?

File: src/ragger/backend/speculos.py
Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:170]: Ok.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:324]: I can remove that... although for the time being this wont be available for other backends...

Comment (xchapron-ledger) [reply src/ragger/navigator/navigator.py:324]: humf, what would be the behavior? You are probably right to keep this for now I guess.
But maybe update the comment to explain the reason?

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:328]: We will either have an unwanted nav instruction or a useless 1 second wait if the screen is already updated to the review flow start.

Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:328]: I'll update according to what you suggest, it won't hurt too much I guess.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:324]: Ok I'll update the TODO

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [reply src/ragger/navigator/navigator.py:328]: not optimal for sure, but I personally think it's better?

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py]: Done.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:327]: Done.

File: .github/workflows/build_and_tests.yml
Comment (xchapron-ledger) [inline .github/workflows/build_and_tests.yml:52]: Do you think this should remain?

File: .github/workflows/build_and_tests.yml
Comment (xchapron-ledger) [inline .github/workflows/build_and_tests.yml:36]: I guess this changes was necessary due to recent changes in app-boilerplate?
Could you put in in another dedicated PR that we could merge quickly?

File: .github/workflows/build_and_tests.yml
Comment (xchapron-ledger) [inline .github/workflows/build_and_tests.yml:26]: To be removed once your speculos Pr is merged I guess?

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:338]:
```suggestion
            # the exception is thrown then we pass and go to the
```
I guess there is a typo in there :)

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:312]: not updated

File: src/ragger/backend/interface.py
Comment (xchapron-ledger) [inline src/ragger/backend/interface.py:337]:
```suggestion
        Wait until the screen content (text) changes compared to what is provided
        by the context parameter. If no context is provided, the function returns
        immediately, returning the current screen context.

        This method may be left void on backends connecting to physical devices,
        where a physical interaction must be performed instead.
        This will prevent the instrumentation to fail (the void method won't
        raise `NotImplementedError`), but the instrumentation flow will probably
        get stuck (on further call to `receive` for instance) until the expected
        action is performed on the device.

        :param timeout: Maximum time to wait for a screen change before an
                        exception is raised.
        :type timeout: float
        :param context: Context to compare screen content with.
        :type context: list
        :return: Latest screen context after the screen change.
        :rtype: list
```
As technically the context is opaque and may contains something else than the text content of the strng?

File: src/ragger/backend/interface.py
Comment (xchapron-ledger) [inline src/ragger/backend/interface.py:318]: Could the type of context be something else than a list? Maybe `Any` or something like that?

**Review by xchapron-ledger — COMMENTED** [review]
Look good to me.
I have added a few minor comments but mostly the real blocker is the merge of the speculos PR :)

File: .github/workflows/build_and_tests.yml
Comment (agrojean-ledger) [reply .github/workflows/build_and_tests.yml:36]: Yes I will open a new PR once @fbeutin-ledger gives me the go. He needs to update the name of binary artifact archive to something more explicit than just "apps"

File: .github/workflows/build_and_tests.yml
Comment (agrojean-ledger) [reply .github/workflows/build_and_tests.yml:26]: Indeed.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:338]: There is. Fixing it right now.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py]: Fixed.

File: .github/workflows/build_and_tests.yml
Comment (agrojean-ledger) [reply .github/workflows/build_and_tests.yml:52]: @xchapron-ledger `pip show protobuf` can be removed but I'm not sure about the downgrade part (`pip install -I protobuf==3.20.0`). The version installed is not compatible with the proto files we have... Don't know when this problem appeared though.

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:318]: `Any` would make everything even more opaque... Not very fond of this idea.

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:337]: See my comment here : https://github.com/LedgerHQ/ragger/pull/50#discussion_r1046954020

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:318]: Actually... If we don't know what the backend returns with `get_screen_content` we should maybe go with what you suggest.

Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Done.

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Done.

File: .github/workflows/build_and_tests.yml
Comment (agrojean-ledger) [reply .github/workflows/build_and_tests.yml:52]: @fbeutin-ledger what do you think ?

File: README.md
Comment (xchapron-ledger) [inline README.md:78]: missing info for `navigate_and_compare_until_text`

File: src/ragger/backend/interface.py
Comment (xchapron-ledger) [inline src/ragger/backend/interface.py:336]:
```suggestion
        :return: Screen context after the screen change.
```

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:361]: I think most of this could be factorize with previous function?
Maybe `navigate_until_text` can redirect on navigate_until_text_and_compare() with empty path and test_case_name or something like that?

**Review by xchapron-ledger — COMMENTED** [review]
Waiting for Speculos Pr to be merge, but definitively I want to use this soon!

File: README.md
Comment (agrojean-ledger) [reply README.md:78]: Adding it now.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:361]: Yes good idea.

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Done.

File: src/ragger/backend/interface.py
Comment (fbeutin-ledger) [inline src/ragger/backend/interface.py:342]: What if we have a screen without text ?

File: src/ragger/navigator/navigator.py
Comment (xchapron-ledger) [inline src/ragger/navigator/navigator.py:438]:
```suggestion
                            timeout: int = 30) -> None:
```

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:359]: Ok.

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:342]: Good question, no idea ^^

File: src/ragger/backend/interface.py
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:359]: Ok.

File: src/ragger/backend/interface.py
Comment (xchapron-ledger) [reply src/ragger/backend/interface.py:358]: It just an opaque value, which has to be passed from one call to another of `wait_for_screen_change`.
the idea is that Speculis backend implementation need a reference to known if the screen has changed.

Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Added an explanation.

File: src/ragger/navigator/navigator.py
Comment (agrojean-ledger) [reply src/ragger/navigator/navigator.py:438]: Ok.

**Comment by lpascal-ledger** [comment]
Will fix issue #45

**Review by xchapron-ledger — DISMISSED** [review]

**Review by fbeutin-ledger — APPROVED** [review]

**Review by fbeutin-ledger — APPROVED** [review]

**Pull request merged**
